SourceX/qol.cpp

```diff
@@ -0,0 +1,460 @@
+/**
+ * @file qol.cpp
+ *
+ * Quality of life features
+ */
+#include "all.h"
+#include "../3rdParty/Storm/Source/storm.h"
+
+DEVILUTION_BEGIN_NAMESPACE
+
+int GetTextWidth(char *s)
+{
+	int l = 0;
+	while (*s) {
+		l += fontkern[fontframe[gbFontTransTbl[(BYTE)*s++]]] + 1;
+	}
+	return l;
+}
+
+int GetConfigIntValue(const char *valuename, int base)
+{
+	if (!SRegLoadValue("devilutionx", valuename, 0, &base)) {
+		SRegSaveValue("devilutionx", valuename, 0, base);
+	}
+	return base;
+}
+
+int highlightItemsMode = 0;
+// 0 = disabled
+// 1 = highlight when alt pressed
+// 2 = hide when alt pressed
+// 3 = always highlight
+bool altPressed = false;
+bool drawXPBar = false;
+bool drawHPBar = false;
+
+void DrawMonsterHealthBar()
+{
+	if (!drawHPBar)
+		return;
+	if (currlevel == 0)
+		return;
+	if (pcursmonst == -1)
+		return;
+	MonsterStruct *mon = &monster[pcursmonst];
+	bool specialMonster = mon->_uniqtype != 0;
+	int currentLife = mon->_mhitpoints;
+	int maxLife = mon->_mmaxhp;
+
+	if (currentLife > maxLife)
+		maxLife = currentLife;
+
+	float FilledPercent = (float)currentLife / (float)maxLife;
+	const int yPos = 180;
+	const int width = 250;
+	const int xPos = (SCREEN_WIDTH) / 2 - BORDER_LEFT;
+	const int height = 25;
+	const int xOffset = 0;
+	const int yOffset = 1;
+	int borderWidth = 2;
+	if (specialMonster)
+		borderWidth = 2;
+	int borderColors[] = { 242 /*undead*/, 232 /*demon*/, 182 /*beast*/ };
+	int borderColor = borderColors[(BYTE)mon->MData->mMonstClass]; //200; // pure golden, unique item style
+	int filledColor = 142;                                          // optimum balance in bright red between dark and light
+	bool fillCorners = true;
+	int square = 10;
+	char *immuText = "IMMU: ";
+	char *resText = "RES: ";
+	char *vulnText = ":VULN";
+	int resSize = 3;
+	int resistColors[] = { 148, 140, 129 }; // { 170,140,129,148,242 };// {168, 216, 200, 242, 142 }; // arcane // fire // lightning // acid
+	WORD immunes[] = { IMUNE_MAGIC, IMUNE_FIRE, IMUNE_LIGHTNING };
+	WORD resists[] = { RESIST_MAGIC, RESIST_FIRE, RESIST_LIGHTNING };
+	WORD mres = mon->mMagicRes;
+
+	int resOffset = 0 + GetTextWidth(resText);
+	for (int k = 0; k < resSize; ++k) {
+		if (!(mres & resists[k]))
+			continue;
+		for (int j = 0; j < square; j++) {
+			for (int i = 0; i < square; i++) {
+				ENG_set_pixel(xPos + i + resOffset, yPos + height + j + yOffset + borderWidth + 2, resistColors[k]);

```

Comment:
Your may also benefit from something like
void ENG_set_line(int x, int size, int y, BYTE col){
//recompute x, y, size with screen constraint
memset(&gpBuffer[sx + BUFFER_WIDTH * sy], size, col)
}
it may remove some inner loops, avoid multiple check and use more agressiv copy

Reply:
well, the square is really tiny ;)